internal/cluster/connection_invocation_handler.go

```diff
@@ -0,0 +1,143 @@
+	bundle.Check()
+	// TODO: make circuit breaker configurable
+	circuitBreaker := cb.NewCircuitBreaker(
+		cb.MaxRetries(3),
+		cb.MaxFailureCount(3),
+		cb.RetryPolicy(func(attempt int) time.Duration {
+			return time.Duration(attempt) * time.Second
+		}))
+	return &ConnectionInvocationHandler{
+		connectionManager: bundle.ConnectionManager,
+		clusterService:    bundle.ClusterService,
+		logger:            bundle.Logger,
+		cb:                circuitBreaker,
+	}
+}
+
+func (h *ConnectionInvocationHandler) Invoke(inv invocation.Invocation) error {
+	_, err := h.cb.Try(func(ctx context.Context) (interface{}, error) {
+		if h.clusterService.SmartRoutingEnabled() {
+			if err := h.invokeSmart(inv); err != nil {
+				h.logger.Warnf("invoking non smart since: %s", err.Error())
+				return nil, h.invokeNonSmart(inv)
+			}
+			return nil, nil
+		} else {
+			return nil, h.invokeNonSmart(inv)
+		}
+	}).Result()
+	return err
+}
+
+func (h *ConnectionInvocationHandler) invokeSmart(inv invocation.Invocation) error {
+	if boundInvocation, ok := inv.(*ConnectionBoundInvocation); ok && boundInvocation.Connection() != nil {
+		return h.sendToConnection(boundInvocation, boundInvocation.Connection())
+	} else if inv.PartitionID() != -1 {
+		if conn := h.connectionManager.GetConnectionForPartition(inv.PartitionID()); conn == nil {
+			return fmt.Errorf("connection for partition ID %d not found", inv.PartitionID())
+		} else {
+			return h.sendToConnection(inv, conn)
+		}
+	} else if inv.Address() != nil {
+		return h.sendToAddress(inv, inv.Address())
+	} else {
+		return h.sendToOwnerAddress(inv)
+	}
+}
+
+func (h *ConnectionInvocationHandler) invokeNonSmart(inv invocation.Invocation) error {
+	if boundInvocation, ok := inv.(*ConnectionBoundInvocation); ok && boundInvocation.Connection() != nil {
+		return h.sendToConnection(boundInvocation, boundInvocation.Connection())
+	}
+	return h.sendToOwnerAddress(inv)
+}
+
+func (h *ConnectionInvocationHandler) sendToConnection(inv invocation.Invocation, conn *Connection) error {
+	if sent := conn.send(inv); !sent {
+		return hzerror.NewHazelcastIOError("packet is not sent", nil)
+	}
+	return nil
+}
+
+func (h *ConnectionInvocationHandler) sendToAddress(inv invocation.Invocation, addr *pubcluster.AddressImpl) error {
+	conn := h.connectionManager.GetConnectionForAddress(addr)
+	if conn == nil {
+		if conn = h.connectionManager.RandomConnection(); conn != nil {
+			h.logger.Trace(func() string {
+				return fmt.Sprintf("address %s not found for invocation, sending to random connection", addr)
+			})
+		} else {
+			h.logger.Trace(func() string {
+				return fmt.Sprintf("sending invocation to %s failed, address not found", addr.String())
+			})
+			return fmt.Errorf("address not found: %s", addr.String())
+		}
+	}
+	return h.sendToConnection(inv, conn)
+}
+
+func (h *ConnectionInvocationHandler) sendToOwnerAddress(inv invocation.Invocation) error {
```

Comment:
There is no definition of owner connection in v4.0. Also, smart invocations should go to a random connection based on the load balancer implementation.

See:
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocation.java#L198
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java#L800

Reply:
Updated at: https://github.com/yuce/hazelcast-go-client/commit/fac635414bfaea419438a7bb8614fb1b2955e45b Currently there's no load balancer apart from random routing.

Reply:
It seems to me that current routing implementation is not random: https://github.com/hazelcast/hazelcast-go-client/pull/552/files#r622204484

Reply:
See: https://github.com/hazelcast/hazelcast-go-client/pull/552#discussion_r622753422